core/src/main/java/org/elasticsearch/common/path/PathTrie.java

```diff
@@ -19,14 +19,45 @@
 
 package org.elasticsearch.common.path;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
 import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
 
 import static java.util.Collections.emptyMap;
 import static java.util.Collections.unmodifiableMap;
 
 public class PathTrie<T> {
 
+    enum TrieMatchingMode {
```

Comment:
> I wonder, if we would always retrieve all matching one (recall optimized) and
> then sort once we have them by the conditions we have here would that be
> simpler code wise?

We currently can't easily do this due to not knowing which handler matched for
the `params`. For example, let's say the second `TrieMatchingMode` matched, but
from the perspective of the caller (`RestController`), it needs both the
`MethodHandlers` object and the parameters that were parsed out of the URI for
the `RestRequest`. It can't use the parameters from the first handler because
they might be a different map that was parsed out.

This is why I used an Iterator with a param supplier, so that they could be
"reset" between invocations, we essentially try to match the request to a
handler multiple times, mutating the `RestRequest` each time and resetting it if
we didn't find an appropriate handler.

I *do* think it would be worth investigating this, however, I think it should be
combined with making `PathTrie` use a builder pattern and `FST` since otherwise
we'll change this whole class around only to have to basically rewrite it for
using an `FST`.

If it's okay with you, I'd like to tackle this in subsequent work that is more
focused, rather than blocking the improve REST PR for internal cleanup. Is that
alright?